repose-aggregator/components/filters/header-normalization/src/main/scala/org/openrepose/filters/headernormalization/HeaderNormalizationFilter.scala

```diff
@@ -0,0 +1,133 @@
+  private var initialized = false
+  private var config: Seq[Target] = _
+  private var metricsMeter: Option[MeterByCategorySum] = _
+
+  override def init(filterConfig: FilterConfig): Unit = {
+    logger.trace("Header Normalization filter initializing...")
+    configurationFile = new FilterConfigHelper(filterConfig).getFilterConfig(DEFAULT_CONFIG)
+
+    logger.info(s"Initializing Header Normalization filter using config $configurationFile")
+    val xsdUrl: URL = getClass.getResource(SCHEMA_FILE_NAME)
+    configurationService.subscribeTo(filterConfig.getFilterName, configurationFile, xsdUrl, this, classOf[HeaderNormalizationConfig])
+
+    metricsMeter = Option(metricsService) match {
+      case Some(service) =>
+        logger.debug("Initializing metrics meter for Header Normalization filter")
+        Some(service.newMeterByCategorySum(classOf[HeaderNormalization], "header-normalization", "Normalization", TimeUnit.SECONDS))
+      case None =>
+        logger.warn("Metrics meter not started for Header Normalization filter due to missing metrics service")
+        None
+    }
+
+    logger.trace("Header Normalization filter initialized.")
+  }
+
+  override def doFilter(servletRequest: ServletRequest, servletResponse: ServletResponse, filterChain: FilterChain): Unit = {
+    val wrappedRequest = new HttpServletRequestWrapper(servletRequest.asInstanceOf[HttpServletRequest])
+
+    config.find { target =>
+      // find the first "target" config element that matches this request (if any)
+      target.url.findFirstIn(wrappedRequest.getRequestURL).isDefined &&
+        (target.methods.contains(wrappedRequest.getMethod) || target.methods.contains(AllHttpMethods))
+    }.foreach { target =>
+      // figure out which headers to remove, and remove them
+      (target.access match {
+        case WhiteList => wrappedRequest.getHeaderNamesScala.diff(target.headers)
+          // written like this to maintain the case-insensitive string comparisons - do not swap
+        case BlackList => target.headers.intersect(wrappedRequest.getHeaderNamesScala)
+      }).foreach(wrappedRequest.removeHeader)
+
+      metricsMeter.foreach(_.mark(s"${target.url}_${wrappedRequest.getMethod}"))
+    }
+
+    filterChain.doFilter(wrappedRequest, servletResponse)
+  }
+
+  override def destroy(): Unit = {
+    logger.trace("Header Normalization filter destroying...")
+    configurationService.unsubscribeFrom(configurationFile, this.asInstanceOf[UpdateListener[_]])
+    logger.trace("Header Normalization filter destroyed.")
+  }
+
+  override def configurationUpdated(config: HeaderNormalizationConfig): Unit = {
+    this.config = config.getHeaderFilters.getTarget.asScala.map { target =>
+      val access = if (target.getBlacklist.isEmpty) WhiteList else BlackList
+      val headers = (access match {
+        case WhiteList => target.getWhitelist
+        case BlackList => target.getBlacklist
+      }).get(0).getHeader.asScala.map(_.getId).toSet
```

Comment:
Side note: JavaConversions provides implicit Java <-> Scala conversions much like JavaConverters does explicitly. It saves you having to type .asScala all the time. This is another thing that doesn't particularly bother me, but knowing about both can be useful. As I've begun to trust implicits more, I've started using JavaConversions myself.